### PR TITLE
Cody prompting guide

### PR DESCRIPTION
**Pull request by chris-sev**
Adding a Cody Prompting Guide that shows best practices for getting the most out of Cody

Contains 2 sections:

- Preparation: Writing good code and comments
- Prompting: How to talk to Cody and bring in specific info + context

### Still to do

- [ ] Add a link to this page somewhere
- [ ] Upload images

### Images

I wasn't sure how to upload images to Google Storage so I'm adding them here. The filenames correspond with the `IMAGE HERE: image-name.jpg` sections in the markdown file.

cody-mention-snippet
cody-mention-file
cody-document-code

### CODE REVIEW

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 1]
```diff
@@ -0,0 +1,199 @@
+# Cody Prompting Guide
+
+In this guide, we'll explore the best practices and tips for effective techniques to maximize your Cody experience.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:3]:
```suggestion
<p className="subtitle">These docs help you learn some of the best practices and tips for effective techniques to maximize your Cody experience.</p>
```

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+## Overarching Recommendation: Treat Cody Like a New Team Member
+
+When collaborating with Sourcegraph Cody, it's helpful to approach it as if you were working with a new engineer who is unfamiliar with your project. Just as you would provide a new team member with detailed information and context about the codebase and the tasks at hand, you should do the same with Cody.
+
+Always lean towards providing more specific information about a certain task to ensure Cody understands your requirements clearly. The more context and details you provide, the better Cody can assist you in generating relevant and accurate code.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:9]:
```suggestion
Always provide more specific information about a task to ensure Cody understands your requirements clearly. The more context and details you provide, the better Cody can assist you in generating relevant and accurate code.
```

Just making the tone more docs friendly. :)

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+✅ Provide a more specific prompt with details:
+
+```
+I have an array of product objects in JavaScript, where each object has the following properties: id, name, category, and price. How can I write a function to filter the products by category?
+```
+
+By treating Cody as a new team member and providing detailed information, you'll establish a more effective collaboration and get better results.
+
+We will split this guide into two parts: Preparation and Prompting.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:27]:
```suggestion
Let's split these docs into two parts: 
```

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+-   Preparation: This part will focus on preparing your code for Cody.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:29]:
```suggestion
-   **Preparation**: How you'll prepare your code for Cody
```
formatting fix

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+-   Prompting: This part will focus on crafting effective prompts for Cody.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:30]:
```suggestion
-   **Prompting**: How you will create effective prompts for Cody
```
formatting fix

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+## Part 1: Preparation
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:32]:
```suggestion
## Preparation
```
We try not to use numbers in headings as these distort the URL pattern :)

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+Before diving into prompting techniques, it's crucial to set the stage for a successful collaboration with Cody. Here are some key preparation steps:
+
+### Choose Descriptive VariNames
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:36]:
```suggestion
### Choose descriptive variable names
```
We use sentence-cased headings unless they have product names.

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+❌ Bad example:
+
+```php
+function calc($i, $t) {
+    // ...
+}
+```
+
+### Write Clear Comments and Docstrings
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:56]:
```suggestion
### Write clear comments and docstrings
```

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+Here is an example of bringing in specific lines of code. The shortcut to do this is to highlight code and press `Opt+L` on Mac or `Ctrl+L` on Windows.
+
+IMAGE HERE: cody-mention-snippet.jpg
+
+Here is an example of `@-` mentioning a specific file:
+
+IMAGE HERE: cody-mention-file.jpg
+
+### Provide Examples and Test Cases
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:128]:
```suggestion
### Provide examples and test cases
```

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+Invalid:
+- "john@example"
+- "jane.doe@example."
+- "invalid.email"
+
+Please write the function
+```
+
+### Iterate and Refine
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:155]:
```suggestion
### Iterate and refine
```

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+Thanks. I forgot to mention:
+
+- The function should take an array of order items as input
+- We need to apply a 10% discount if the total price exceeds $100
+- Final total should be rounded to two decimal places
+
+Please update the function
+```
+
+### Leverage Cody's Strengths
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:177]:
```suggestion
### Leverage Cody's strengths
```

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+Utilize Cody's capabilities for generating boilerplate code, common patterns, and repetitive tasks. Prompt it to assist with unit tests, documentation, and error handling to save time and ensure code quality.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:179]:
```suggestion
Utilize [Cody's capabilities](/cody/capabilities) for generating boilerplate code, common patterns, and repetitive tasks. Prompt it to assist with unit tests, documentation, and error handling to save time and ensure code quality.
```
let's link to our Capabilities docs

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+- Empty ratings array should return 0
+- Array with one rating should return that rating
+- Array with multiple ratings should calculate the average correctly
+
+Make sure the tests cover any edge cases or potential issues. Thanks!
+```
+
+You can also use the **Generate Tests** command to generate tests for your code.
+
+## Conclusion
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:195]:
```suggestion
## Conclusion
```
I'd suggest let's remove this conclusion part since it does not align with the docs content type. :)

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+
+By following these best practices for preparation and prompting, you can effectively collaborate with Sourcegraph Cody and leverage its AI capabilities to boost your coding productivity. Remember, treating Cody like a new team member and providing detailed information is key to a successful collaboration. By following these best practices and examples, you'll be able to leverage Cody's AI capabilities effectively and streamline your coding workflow.
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:197]:
```suggestion
It's recommended to try out these tips and make your experience better with Cody. If you have any questions regarding Cody, you can always [ask our community](https://community.sourcegraph.com), [Discord](https://discord.com/invite/s2qDtYGnAE), or [create a post on X](https://twitter.com/sourcegraphcody).
```
Instead, we can add something like this. :)

File: docs/cody/core-concepts/prompting-guide.mdx [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+
+Happy coding with Cody!
```
Comment (MaedahBatool) [inline docs/cody/core-concepts/prompting-guide.mdx:199]:
```suggestion

```

**Review by MaedahBatool — CHANGES_REQUESTED** [review]
Hey @chris-sev thanks a ton for contributing these docs. Left some minor suggestions. Let me know your thoughts about it. :)

**Comment by MaedahBatool** [comment]
hey @chris-sev catching up on this PR? Are we still doing this?
